go/internal/otel/otel.go

```diff
@@ -0,0 +1,374 @@
+package otel
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/log"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/propagation"
+	sdklog "go.opentelemetry.io/otel/sdk/log"
+	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/resource"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.4.0"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/launchdarkly/observability-sdk/go/internal/logging"
+	"github.com/launchdarkly/observability-sdk/go/internal/metadata"
+)
+
+type providerInstances struct {
+	tracerProvider trace.TracerProvider
+	loggerProvider log.LoggerProvider
+	meterProvider  metric.MeterProvider
+}
+
+type otelInstances struct {
+	tracer trace.Tracer
+	logger log.Logger
+	meter  metric.Meter
+}
+
+// Config contains the configuration for the OTLP provider.
+type Config struct {
+	OtlpEndpoint       string
+	ResourceAttributes []attribute.KeyValue
+	Sampler            sdktrace.Sampler
+}
+
+func defaultInstancesValue() *atomic.Value {
+	v := &atomic.Value{}
+	providers := getDefaultProviders()
+	instances := &otelInstances{
+		tracer: providers.tracerProvider.Tracer(
+			metadata.InstrumentationName,
+			trace.WithInstrumentationVersion(metadata.InstrumentationVersion),
+			trace.WithSchemaURL(semconv.SchemaURL),
+		),
+		logger: providers.loggerProvider.Logger(
+			metadata.InstrumentationName,
+			log.WithInstrumentationVersion(metadata.InstrumentationVersion),
+			log.WithSchemaURL(semconv.SchemaURL),
+		),
+		meter: providers.meterProvider.Meter(
+			metadata.InstrumentationName,
+			metric.WithInstrumentationVersion(metadata.InstrumentationVersion),
+			metric.WithSchemaURL(semconv.SchemaURL),
+		),
+	}
+	v.Store(instances)
+	return v
+}
+
+// writeLock is used to synchronize access to the instances and config atomic values.
+// Starting or stopping OTLP should be done with this lock held.
+// Access to the OTLP instances should be done with the instances atomic value and
+// does not require holding the lock.
+//
+//nolint:gochecknoglobals
+var writeLock sync.Mutex
+
+// Instances are stores in an atomic so that they can have consistent visibility
+// within goroutines without using locks. If this was not an atomic then goroutines
+// which start before the OTLP configuration was complete may never have visibility
+// of the current values.
+//
+// The similar otel functions, such as otel.Tracer, also use atomics to store the
+// current values.
+//
+//nolint:gochecknoglobals
+var instances *atomic.Value = defaultInstancesValue()
```

Comment:
The comment contains an explanation, but a good amount of globals were not strictly thread-safe in the previous implementation. It likely would not matter in day-to-day unless you were starting the SDK after lots of other go routines were running.